src/goto-programs/static_analysis.cpp

```diff
@@ -52,6 +52,75 @@ void static_analysis_baset::operator()(const goto_programt &goto_program)
   fixedpoint(goto_program, goto_functions);
 }
 
+void static_analysis_baset::get_concurrent_idts(
+  const goto_functionst &goto_functions,
+  std::set<irep_idt> &idts)
+{
+  idts.insert(pthread_set.begin(), pthread_set.end());
+  // We use a recursive approach here to handle this situation:
+  // func1 {func2(); func3()} we need to handle the inner function
+  for(const auto &identifier : pthread_set)
+  {
+    insert_idt(goto_functions, idts, identifier);
+  }
+}
+
+void static_analysis_baset::insert_idt(
+  const goto_functionst &goto_functions,
+  std::set<irep_idt> &idts,
+  const irep_idt &idt)
+{
+  goto_functionst::function_mapt::const_iterator it =
+    goto_functions.function_map.find(idt);
+
+  const goto_functiont &goto_function = it->second;
+
+  working_sett working;
+
+  put_in_working_set(working, goto_function.body.instructions.begin());
+
+  while(!working.empty())
+  {
+    locationt l = get_next(working);
+    visit_function(l, working, goto_function.body, goto_functions, idts);
+  }
+}
+
+void static_analysis_baset::visit_function(
+  locationt l,
+  working_sett &working,
+  const goto_programt &goto_program,
+  const goto_functionst &goto_functions,
+  std::set<irep_idt> &idts)
+{
+  goto_programt::const_targetst successors;
+
+  goto_program.get_successors(l, successors);
+
+  for(goto_programt::const_targetst::const_iterator it = successors.begin();
+      it != successors.end();
+      it++)
+  {
+    locationt to_l = *it;
+    // Recursively fetches the function calls inside and adds them to the set,
+    // func1{func2}->func2{func3}->fun3(),
+    // add idts of func1, func2, func3.
+    if(l->is_function_call())
+    {
+      const code_function_call2t &code = to_code_function_call2t(l->code);
+      irep_idt identifier = to_symbol2t(code.function).get_symbol_name();
+
+      // We should ignore all function calls from libraries
```

Comment:
Explain why we must ignore all function calls from libraries. Why do we consider only `pthread'?

Reply:
Here we are using a recursive approach, if there is a function call inside, we find the function and double check if there is a function call inside, In fact, data races do not occur in these libraries, and for some libraries, infinite recursion may occur. There's no good way to identify all of them, so for now we'll only consider the following:
```
pthread_mutex_lock( )
pthread_mutex_unlock( )
pthread_exit(NULL)
```

Reply:
Thanks for this explanation. Can you improve the documentation in the code?